sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADIssuerJWSKeySelector.java

```diff
@@ -0,0 +1,78 @@
+import com.nimbusds.jose.KeySourceException;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.jwk.source.RemoteJWKSet;
+import com.nimbusds.jose.proc.JWSAlgorithmFamilyJWSKeySelector;
+import com.nimbusds.jose.proc.JWSKeySelector;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jose.util.DefaultResourceRetriever;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.proc.JWTClaimsSetAwareJWSKeySelector;
+import java.net.URL;
+import java.security.Key;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Selecting key candidates for processing a signed JWT which provides access to the JWT claims set in addition to the
+ * JWS header.
+ */
+public class AADIssuerJWSKeySelector implements JWTClaimsSetAwareJWSKeySelector<SecurityContext> {
+
+    private AADTrustedIssuerRepository trustedIssuerRepository;
+
+    private int connectTimeout;
+
+    private int readTimeout;
+
+    private int sizeLimit;
+
+    private final Map<String, JWSKeySelector<SecurityContext>> selectors = new ConcurrentHashMap<>();
+
+    public AADIssuerJWSKeySelector(AADTrustedIssuerRepository trustedIssuerRepository, int connectTimeout,
+        int readTimeout, int sizeLimit) {
+        this.trustedIssuerRepository = trustedIssuerRepository;
+        this.connectTimeout = connectTimeout;
+        this.readTimeout = readTimeout;
+        this.sizeLimit = sizeLimit;
+    }
+
+    @Override
+    public List<? extends Key> selectKeys(JWSHeader header, JWTClaimsSet claimsSet, SecurityContext context)
+        throws KeySourceException {
+        String iss = (String) claimsSet.getClaim(AADTokenClaim.ISS);
+        if (trustedIssuerRepository.getTrustedIssuers().contains(iss)) {
+            return this.selectors.computeIfAbsent(iss, this::fromIssuer).selectJWSKeys(header, context);
+        }
+        throw new IllegalArgumentException(
+            "The current issuer is not included in the trusted issuers, no jws key selector is configured.");
+    }
+
+    private JWSKeySelector<SecurityContext> fromIssuer(String metadataUrlPrefix) {
+        return Optional.ofNullable(metadataUrlPrefix)
+            .map(AADJwtDecoderProviderConfiguration::getConfigurationForOidcIssuerLocation)
+            .map(this::fromUri)
+            .orElseThrow(() -> new IllegalArgumentException("Cannot create JWSKeySelector."));
+    }
+
+    private JWSKeySelector<SecurityContext> fromUri(Map<String, Object> configuration) {
+        String uri = configuration.get("jwks_uri").toString();
+        try {
+            DefaultResourceRetriever jwkSetRetriever = new DefaultResourceRetriever(connectTimeout, readTimeout,
+                sizeLimit);
+            JWKSource<SecurityContext> jwkSource = new RemoteJWKSet<>(new URL(uri), jwkSetRetriever);
+            return JWSAlgorithmFamilyJWSKeySelector.fromJWKSource(jwkSource);
+        } catch (Exception ex) {
```

Comment:
There's no need to catch the exception here. we should add the throw to the function.

Reply:
Seems not easy to implement since lambda can't throw checked exception.